### PR TITLE
update of contributing guide based on team meeting

### PR DESCRIPTION
**Pull request by mejackreed**
This pull request summarizes agreed upon contributing process for the Developer Playbook on 5/21/2015.

 - Adds helpful links for GitHub issues, pull requests, labels, and emoji
 - Splits into 3 sections for readability

### CODE REVIEW

**Comment by cbeer** [comment]
:+1: I see a couple typos ("Proposals are be") and some quibbles that I'm happy to address in future PRs.

**Comment by azaroth42** [comment]
:+1:

On Thu, May 21, 2015 at 8:33 PM, Chris Beer <notifications@github.com>
wrote:

> [image: :+1:] I see a couple typos ("Proposals are be") and some quibbles
> that I'm happy to address in future PRs.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/sul-dlss/DeveloperPlaybook/pull/21#issuecomment-104492226>
> .
>



-- 
Rob Sanderson
Information Standards Advocate
Digital Library Systems and Services
Stanford, CA 94305

**Comment by darrenleeweber** [comment]
:-1:  As it stands, this is a recipe for group think.  I can't accept these rules unless each issue and PR can reference authorities in respected communities outside of DLSS and these authorities are clearly recognized as reasonable sources for best practices.

**Comment by ndushay** [comment]
:+1:   @mejackreed   "Consensus" is the correct spelling, btw.    Eventually we should probably say something more than "don't be a jerk" in that section.

**Comment by mejackreed** [comment]
@ndushay thanks for the spellcheck! :joy_cat:

**Comment by cbeer** [comment]
I actually believe group-think is part of the goal of the playbook, and do not anticipate it being a problem. Hopefully, our consensus and existing practices are driven by practices in the wider communities, and when they aren't, I believe we're all reasonable people who can respond and adapt to change.

We're not setting practices in stone, we're just cataloging current consensus and expectations of each other. The playbook also does not require that projects follow it unconditionally, just that there is a default set of expectations, and strong encouragement for applications to follow them or document their deviations from existing practice.

**Comment by eefahy** [comment]
@darrenleeweber I don't agree that every issue or PR would need to be externally resourced in order to avoid 'group think' and to prove both of our points please read the Consensus section from: http://web.mit.edu/publicdisputes/practice/cbh_ch1.html

**Comment by azaroth42** [comment]
@darrenleeweber I understand your point, but as @cbeer said, the goal of generating the playbook at the moment is to describe current practices rather than establish new ones.  We want to ensure that there is consensus around the topics that affect us day to day, and discover the areas that we do not have consensus on so we can try to come to general agreement.

Consensus currently being at least four people agree and no more than one disagrees. Disagreement being, as you stated, "I can't accept that." or similar levels of unhappiness.

**Comment by darrenleeweber** [comment]
WRT the consensus section of http://web.mit.edu/publicdisputes/practice/cbh_ch1.html (at the risk of misinterpreting that document because I want to take the time to read all of it and I've only read the paragraphs about consensus); I agree with the intention to consider all interests and strive for unanimous agreement.  As it stands, the policy for issue and PR in the playbook do not satisfy my idea of striving for consensus.  The playbook definition appears to be only 1 week for commentary, acceptance after only 4 thumbs up and *dismissing* any 1 thumbs down.  I would prefer a more inclusive voting and discussion process.  I would like to see a voting process that (a) identifies all the parties in DLSS (developers, dev-ops, sys-admins, managers) and (b) clearly identifies majority acceptance while trying to address concerns of any minority views.  At present, I'm concerned that contributions to this playbook are coming from a minority of the DLSS staff.

Also, it's important that anything we call a 'best practice' is also a reasonable practice in the wider OSS community.  For example, deviating from the puppet upgrade process and working with too many in-house customizations, even if they were an in-house consensus, has led to a mine field and that is now getting untangled.

**Comment by azaroth42** [comment]
The process is to ensure that we make progress rather than 
 * getting stuck in the weeds trying to gain consensus when the point is to catalog what we already have consensus on
 * getting stuck waiting for people to vote on things that don't concern them at all, or they have no strong opinions on
 * getting stuck because the process drags on for too long and everyone loses interest

Agree that "best practice" is NOT the correct description of what we're trying to get to here. We're trying to collect current working practices and expectations, which can be improved iteratively. I'll submit a PR to that effect.

**Comment by darrenleeweber** [comment]
I have to admit that this process is leading me to question the meanings of 'best practice', 'consensus' and 'progress'.

**Comment by azaroth42** [comment]
Given that until this PR hits there *is* no process... I propose we merge it on the grounds that it accurately reflects the consensus from yesterday's meeting.

**Comment by drh-stanford** [comment]
:+1:  this looks like an accurate representation from the meeting

**Comment by cbeer** [comment]
@darrenleeweber with the discussions that have happened on other tickets, are you still :-1: and completely against this proposal?

**Comment by darrenleeweber** [comment]
@cbeer et al. - there is a fine line drawn between making progress on documenting recommended practices and arriving at consensus or a reasonable majority, all in the context of optional voting.  Given the size of DLSS and the actual staff who care about this, the current rules may be useful, but they can allow the imposition of practices that are actually agreed upon by a small group and not necessarily representative of DLSS as a whole.  I've aired my concerns about the dangers of group think and my concerns about referring to outside authorities on important practices; the latter is one way to ameliorate group think tendencies.  Given that the latter has not been well received, my concerns remain and I don't want to accept this as is.  The addition of general goals has helped, in that, it states these practices are not retroactive and they are reasonable guidelines rather than obligations (although my concern about this is that the latter might actually be closer to the true intent and there may be significant resistance to new projects that deviate, even when alternate solutions are well reasoned, well documented, and properly implemented, tested, etc.).

Part of the sub-text here is the consolidation of dev-ops resources and experience on familiar platforms and practices, esp. where that makes it easier to work with puppet and other deployment systems.  In the past year, I've had a hard time with wanting to work with a triple store and mongodb, for example, because the mantra is that dev-ops will not support it in prod (it was suggested that if I want to use it prod, I will have to be the admin support for it).  As an engineer, I don't want dev-ops to tell me how to architect a solution; I want dev-ops to support or learn to support the solution.  If this playbook becomes the gospel that dev-ops can refer to so they can justify limiting the solutions for engineers, then I am dead against it.

**Comment by azaroth42** [comment]
Darren, lets discuss face to face after auth discussion this afternoon.

**Comment by cbeer** [comment]
@azaroth42 @darrenleeweber did your face-to-face conversation happen? I saw a couple new tickets come through yesterday, and maybe they address @darrenleeweber's concerns about this proposal, at least enough to turn :-1: into -0 with reservations?

I hear your concern about a vocal minority pushing an agenda, but, to my mind, there are a number of safeguards built into this proposed process and into the ethos of the playbook as a whole, including:

- presumably all are acting in good faith, and the issues we bring to the playbook are tools, approaches, or methodologies that we believe improve the quality of our work.
- these are living guides, and all are welcome to contribute new ideas or modify existing practices. Even if a proposal is accepted, there's always room for disagreement and modification.
- :+1:s, and the "voting" process are less about majority (or plurality, or those-who-show-up) win, but an easy way to quantify support for a proposal, and give us a way to know when there is "enough" support to merge a proposal. One of the downsides to our previous lack-of-process was several proposals with significant support, but people unsure when and how to merge something.
- similarly, :-1:s, which we should never treat lightly. If we have disagreement within the group about an issue, we should work to resolve it and improve the proposal to accurately reflect the needs of the group. Also, the review procedure offered here provides a mechanism for a small minority to ensure their views are heard and addressed.

The process here is far from perfect, but it, like the practices we document, can evolve as we begin to apply it.

**Comment by darrenleeweber** [comment]
Perhaps we can split this PR into (a) guidelines on submitting proposals and (b) rules for accepting proposals.  I can :+1: for (a), but somewhat on the fence about (b) and inclined toward :-1: on it.

**Comment by cbeer** [comment]
@darrenleeweber is it possible you could live this proposal as-is, and anticipate modifying the process as we apply the process and learn how to do it better? If not, could you give an indication of the process you envision? Maybe there's a way we can adapt what's here to meet all our needs.

And, for the record, I'm personally not thrilled with the language of this proposal, which feel too close to voting for my comfort. That said, it seems to have general support and is something I can live with for now.

**Comment by atz** [comment]
@darrenleeweber, by a literal reading, you are arguing that we cannot agree to *anything* without adopting it from another community.  This is so self-limiting that I must interpret it as rhetorical.  In particular, there are points that we *already* agree on, without appeal to such authority.  Also there are certain areas where this group *is* the authority (e.g. aspects of hydra & blacklight development) or at least peer to other groups in the broader community.  While they are useful for depth and context, requiring external references would only shift the debate to what sources were authoritative and for what topics, since the citation itself does not resolve anything.  And of course sources differ: the universe is not uniform, does not agree.  

I am sympathetic to the concern that this part (b) is different than the other proposed parts of the Playbook inasmuch as it is self-referential/meta/governing.  In that sense it is the most critical part.  But I see no devops subtext.  There is no need for a hypothetical document to consolidate devops resources because there are so few of them they are *already* effectively consolidated.  Quite to the contrary, this Playbook/process would be an effective vehicle for promoting a platform to an unaware or even oppositional team member, given that there was consensus on the platform.  Obviously, no such consensus exists re: mongo.  NOSQL stores are something many of us have experience with and many regard with interest.  Having or not having this document doesn't change that.

I agree w/ @cbeer that we should avoid language that approaches voting and vote counting.  In particular there is nothing magic about the numbers.  We should be explicit that committee-style formal voting is exactly **not** the type of system we are trying to build.

**Comment by cbeer** [comment]
> We should be explicit that committee-style formal voting is exactly not the type of system we are trying to build.

@atz is that an important requirement before we merge this, or an enhancement after-the-fact? I'm happy either way, but inclined to ship it and ponder the right context, language, and process later.

**Comment by azaroth42** [comment]
To comment on the "voting" part...  I completely agree that the +1s / -1s are not votes in any meaningful sense of that word.

My opinion is that we are using github's +1 / -1 as a way of **gauging the current feeling** about a particular topic in an **asynchronous** environment in such a way as to be inclusive as possible without blocking based on non-response or having to gain unanimous agreement which **will not work without face to face discussion**.  If we could sit around for 2 hours a week with lunch in 370, we wouldn't need this process **at all**. However that process would be very expensive, and lunch would be the smallest cost, and one that I was happy to bear for a kickoff meeting.  I had hoped that starting off in a positive way with significant progress (IMHO) would provide sufficient momentum to get us past the difficult uphill climb we're in at the moment.

Now, if we think this is important, we could ask for more time to work on it as a department in the consensus building spirit of the first meeting.  If we think that would be a net gain, I'm happy to champion it and we can ditch the whole asynchronicity.

**Comment by ndushay** [comment]
I'm still for "merge" on this PR, and tweak the text later.  I think the relevant question to this PR is what @cbeer asked:

> @darrenleeweber is it possible you could live this proposal as-is, and anticipate modifying the process as we apply the process and learn how to do it better? If not, could you give an indication of the process you envision? Maybe there's a way we can adapt what's here to meet all our needs.

and I also think the most relevant response to this issue is from @drh-stanford:

>  this looks like an accurate representation from the meeting

If we need to re-visit our process, then let's do that f2f, and then we can amend this PR's text *after* that f2f discussion.

I'm trying to think this way right now, for the issues and PRs before we meet f2f again:
* is the suggested idea / text an improvement over nothing (since we are in the initial phase)
* can i live with it as is?
* we will be able to revise anything in the future, subject to our process
* we are trying out a process - it may need revision
* we are doing something new - mistakes are allowed

**Comment by darrenleeweber** [comment]
References to something like

>  this looks like an accurate representation from the meeting

are important to the point of this discussion with regard to acceptance.  I, for one, was not present at the meeting so I know first hand how it feels to have a group wanting to accept something based on the meeting when I don't agree with it and I had no opportunity to discuss it at the meeting.  This playbook is an async process that can be enhanced or facilitated by F2F meetings, but dissent from someone who was not present at a meeting cannot be dismissed simply because they did not participate in the meeting and a majority or consensus was arrived at in the meeting.

I can't follow the logic of everything proposed in comments from @atz, but I get some of the gist.  I'm not wanting to disparage or in any way cast doubt on the expertise of the DLSS staff.  I'm simply concerned that the acceptance criteria, as is, do not provide a reasonable process.

As it stands, I don't feel comfortable accepting it.  I see a lot of value in the PR, so it's difficult to hold out on a point about the acceptance of contributions.  Given some additional concerns or consideration about 'voting', it might be better to split this into (a) ways to contribute or how to contribute and (b) process for discussion, review, and accepting or rejecting contributions.

**Comment by cbeer** [comment]
> Given some additional concerns or consideration about 'voting', it might be better to split this into (a) ways to contribute or how to contribute and (b) process for discussion, review, and accepting or rejecting contributions.

@darrenleeweber you said that in your last comment as well, and I would again ask you indicate something that you /would/ find acceptable, so we can try to reach some kind of agreement.

**Comment by darrenleeweber** [comment]
In the light of my concerns about avoiding groupthink patterns and wanting to refer to outside information, without wanting to be identified as an authority on the topic myself, I'm willing to take some time to review additional content, such as:
  - http://www.ryanhartwig.com/9-strategies-to-avoid-groupthink/

**Comment by cbeer** [comment]
@darrenleeweber thanks for that link. How do you suggest we apply the strategies to this particular situation?

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -1,6 +1,34 @@
...
+   - What is the danger if there isn't a best practice?
+   - How do people most effectively follow it?
+  
+### Gaining Consensus
+ - Consensus is gauged by people commenting with :+1: and :-1: ([GitHub Emoji Cheatsheet](http://www.emoji-cheat-sheet.com/))
+ - :-1: must be described concretely as to why there is disagreement with that specific proposal
+ - After one week, if there are less than four +1s or more than one -1, and no ongoing clarification discussion, the proposal is set aside for in person discussion.
+ - Every second week we have a 1 hour long meeting for discussing the proposals that have been set aside.  Proposals get no more than 15 minutes discussion time.
+
+### Accepting a Proposal
```
Comment (darrenleeweber) [inline CONTRIBUTING.md]: Drop this section for a later PR.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -1,6 +1,34 @@
...
+ - Proposals are made by creating a GitHub [issue](https://help.github.com/articles/creating-an-issue/) or [pull request](https://help.github.com/articles/creating-a-pull-request/) (PR).
+ - Issues containing the proposal are tagged with the area(s) under which they fall using [GitHub labels](https://github.com/sul-dlss/DeveloperPlaybook/labels).
+ - The proposal is an appropriate granularity requirement to capture existing consensus.
+ - The proposal is of the form When <doing activity>, we [MUST|SHOULD] <use described approach>.
+ - Proposals are to be scoped to whether it applies universally, or can be decided on by individual project teams.
+ - The body of the issue must describe the rationale for why the entry is needed
+   - What is the danger if there isn't a best practice?
+   - How do people most effectively follow it?
+  
+### Gaining Consensus
```
Comment (darrenleeweber) [inline CONTRIBUTING.md]: Revise this section, let's consider healthy team communication practices, e.g.
  -  http://www.ryanhartwig.com/9-strategies-to-avoid-groupthink/
  -  http://www.ryanhartwig.com/6-strategies-to-stir-positive-conflict/
For example, discussions should encourage divergent views and healthy, respectful conflict in the spirit of open, honest communication.  Striving for consensus at the cost of avoiding healthy conflict could arrive at decisions that ultimately lead to a lack of effective team commitment.

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -1,6 +1,34 @@
...
+
+### Accepting a Proposal
+ - If at the end of the week or in person discussion there is four or more +1s and one or fewer -1, then the proposal is accepted.
+ - After a proposal is accepted the proposer writes it up in markdown and submits a PR to the playbook, tagging the issue (if it’s not already a pull request).
+ - PRs must not introduce new decisions beyond what was discussed in the issue.  If further proposals are needed to flesh out the PR's content, they must be submitted using the method above.
+ - Consensus review and accept the PR or work with the proposer to fix any issues. Fourth person to come along and :+1: ships it.
+ - Otherwise if the proposal is not accepted it gets tagged "defer" as a note that consensus could not be reached and closed.
+ - Deferred proposals should have consensus building around them before being reintroduced.
+
+## Guiding Principles
```
Comment (darrenleeweber) [inline CONTRIBUTING.md]: Drop this section for a later PR.

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -1,6 +1,34 @@
...
+
+ - Don’t be a jerk :smiley:
 
 ## Getting Feedback
```
Comment (darrenleeweber) [inline CONTRIBUTING.md]: Drop this section for a later PR.

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -1,6 +1,34 @@
 # Contributing
 
-We love pull requests. If you have something you want to add or remove, please open a new pull request.
+If you have something you want to add or remove, please use the following agreed upon process.
+
+### Initial Proposal Creation
+ - Proposals are made by creating a GitHub [issue](https://help.github.com/articles/creating-an-issue/) or [pull request](https://help.github.com/articles/creating-a-pull-request/) (PR).
+ - Issues containing the proposal are tagged with the area(s) under which they fall using [GitHub labels](https://github.com/sul-dlss/DeveloperPlaybook/labels).
+ - The proposal is an appropriate granularity requirement to capture existing consensus.
```
Comment (darrenleeweber) [inline CONTRIBUTING.md]: Proposals might be contributed with a view to raising discussion without necessarily having any prior concensus on the issue.

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -1,6 +1,34 @@
...
+ - The proposal is of the form When <doing activity>, we [MUST|SHOULD] <use described approach>.
```
Comment (darrenleeweber) [inline CONTRIBUTING.md]: In the light of proposals that do not necessarily already capture consensus, the terms 'MUST' and 'SHOULD' imply obligations that are not consistent with the goals of the playbook to document accepted practices, without constraining healthy creativity that may diverge from the current practices (for good reasons).

File: CONTRIBUTING.md (hunk 2 above)
Comment (cbeer) [inline CONTRIBUTING.md]: Can you be more specific about how we can do that? Clearly we're already having a vibrant discussion for the issues that have been raised, so I'm not clear what, in specific, you'd like us to do differently.

Comment (darrenleeweber) [inline CONTRIBUTING.md]: Yes, that's a topic of another PR and I too am not entirely clear about what form that takes, but my view is that it is not what we have now in this section.

File: CONTRIBUTING.md (hunk 6 above)
Comment (cbeer) [inline CONTRIBUTING.md]: I don't think that was necessarily the intent of that statement (although, looking at the issues raised so far, we've already deviated from that practice). I think the goal was to provide a strong statement of what problem a proposal is designed to address, e.g.

- "When developing projects, we must use version control"
- "When releasing code, we must apply a license"
- "When documenting best practices, we should link to examples of those practices being used"

Further, in https://github.com/sul-dlss/DeveloperPlaybook/pull/29/files (which, I guess, is blocked by this?), I believe we explicitly address your concern (perhaps prompted by your concerns elsewhere?):

> The goals are explicitly NOT to [...] Mandate or set in stone any unbreakable requirements. They are expectations that can be deviated from for good, documented reasons.

File: CONTRIBUTING.md (hunk 2 above)
Comment (cbeer) [inline CONTRIBUTING.md]: That's fine. Let's work on that pull request. I'm happy to help if you could suggest a broad outline or even specific elements that should be incorporated. As is, I feel like I'm in the dark and find it difficult to contribute towards something we can agree on.

**Comment by atz** [comment]
:+1:  Given the agreement on intent and un-voting-ness, yeah, I'm comfortable adjusting parts later.

**Comment by cbeer** [comment]
Merging this per our face to face meeting last week, where there was general agreement that we would iterate on this process going forward.

**Pull request merged**